crates/apub/src/http/inbox.rs

```diff
@@ -0,0 +1,208 @@
+use crate::{
+  activities::WithPublished,
+  activity_lists::SharedInboxActivities,
+  fetcher::user_or_community::UserOrCommunity,
+};
+use activitypub_federation::{
+  actix_web::inbox::{receive_activity, receive_activity_parts},
+  config::Data,
+};
+use actix_web::{http::header::HeaderMap, web::Bytes, HttpRequest, HttpResponse};
+use chrono::{DateTime, Local, TimeDelta, Utc};
+use http::{Method, Uri};
+use lemmy_api_common::context::LemmyContext;
+use lemmy_db_schema::{newtypes::InstanceId, source::instance::Instance};
+use lemmy_utils::error::LemmyResult;
+use once_cell::sync::Lazy;
+use rand::seq::IteratorRandom;
+use serde::Deserialize;
+use std::{
+  cmp::Ordering,
+  collections::{BinaryHeap, HashMap},
+  sync::{Arc, RwLock},
+  thread::available_parallelism,
+  time::Duration,
+};
+use tokio::{spawn, task::JoinHandle, time::sleep};
+use tracing::info;
+use url::Url;
+
+/// Handle incoming activities.
+pub async fn shared_inbox(
+  request: HttpRequest,
+  bytes: Bytes,
+  data: Data<LemmyContext>,
+) -> LemmyResult<HttpResponse> {
+  match serde_json::from_slice::<WithPublished<()>>(&bytes)?.published {
+    Some(published) => {
+      // includes published timestamp, insert to queue to ensure that activities are processed
+      // in correct order even when delivered out of order.
+      let request_parts = (
+        request.headers().clone(),
+        request.method().clone(),
+        request.uri().clone(),
+      );
+
+      #[derive(Deserialize)]
+      struct Id {
+        id: Url,
+      }
+
+      let activity_id = serde_json::from_slice::<Id>(&bytes)?.id;
+      let domain = activity_id.domain().unwrap().to_string();
+      let instance = Instance::read_or_create(&mut data.pool(), domain)
+        .await
+        .unwrap();
+
+      let mut lock = ACTIVITY_QUEUE.write().unwrap();
+      let instance_queue = lock.entry(instance.id).or_insert(BinaryHeap::new());
+      while instance_queue.len() > 5 {
+        // TODO: must not hold lock here
+        sleep(Duration::from_millis(100)).await;
+      }
+      instance_queue.push(InboxActivity {
+        request_parts,
+        bytes,
+        published,
```

Comment:
This wont work well if `published` is in the future, eg due to out of sync clocks. Should probably handle the activity immediately in that case.